libherokubuildpack/src/buildpack_output/util.rs

```diff
@@ -0,0 +1,183 @@
+use std::fmt::Debug;
+use std::io::Write;
+
+#[cfg(test)]
+use std::sync::{Arc, Mutex};
+
+pub(crate) fn prefix_first_rest_lines(
+    first_prefix: &str,
+    rest_prefix: &str,
+    contents: &str,
+) -> String {
+    let first_prefix = String::from(first_prefix);
+    let rest_prefix = String::from(rest_prefix);
+
+    prefix_lines(contents, move |index, _| {
+        if index == 0 {
+            first_prefix.clone()
+        } else {
+            rest_prefix.clone()
+        }
+    })
+}
+
+pub(crate) fn prefix_lines<F: Fn(usize, &str) -> String>(contents: &str, f: F) -> String {
+    use std::fmt::Write;
+
+    let lines = contents.split_inclusive('\n').enumerate().fold(
+        String::new(),
+        |mut acc, (line_index, line)| {
+            let prefix = f(line_index, line);
+            let _ = write!(acc, "{prefix}{line}");
+            acc
+        },
+    );
+
+    if lines.is_empty() {
+        f(0, "")
+    } else {
+        lines
+    }
```

Comment:
> Re the intended behaviour, I'm curious why adding a prefix is still desirable when passed an empty str? "Blank" lines will actually have a contents of "\n" so this special casing isn't needed for them. ie: If I remove the special-casing here, the tests in the build_output/mod.rs still pass?

If someone calls step or section I think it would be weirder to see:

```

  - Seemingly floating bullet
```

Than this:

```
-
  - Sub-step under an empty bullet above
```

It's an exercise in determining the "least wrong" thing to do. I think showing the bullet point with nothing behind it at least gives a hint to the programmer that they emitted an empty output as opposed to "where on earth did this newline come from.

Reply:
> If someone calls step or section

I think there are some words missing here - did this mean "If someone calls step or section with an empty string argument"?

eg: `.section("")`

If so, I agree it improves the UX in that scenario :-)

Would you mind adding a small comment explaining why this conditional is needed?

Reply:
I pushed a commit. I refactored to make it clearer that the split_inclusive might not yield anything. Then I added tests and some docs for all these methods.

Reply:
Thank you for adding the test and comments in ab5d9a5579b3d9b1a7b54da80f27c9134f0d9b3b.

I see that commit also adjusted the implementation, with the reasoning of:

> I think it makes more logical sense to encode that information in the code, so I refactored

I absolutely agree in the general case that it makes sense to think about adjusting the code first, then documenting after, since clearer code can mean it becomes self-documenting.

However, in this specific case I found the code harder to understand (not from "what" it does, but the "why") after the refactor than it was before.

I think this is highlighted most clearly by looking at the rustdocs that added to explain what's happening:

> If an empty string is provided, ...

You'll note that they describe it as an "if an empty string" (which matches perfectly the conditional that was used before), and not "if there are no lines after splitting" (which is what the new implementation suggests). I know both implementations are equivalent functionality wise, but I think if the rustdocs are clearer when they use the former wording, then the code is too (and it's clearer when both the code and the rustdocs match in their explanation as it were).

I spoke with Manuel about this and said the previous implementation was clearer to him too.

As such in the interests of saving time here, I've reverted the implementation change (but left the docs and tests) in:
7d8fa16e84febe240ad2dfa4eb9587e5ee5666b0

Reply:
(Feel free to mark this as resolved when you've read it; leaving it open for now otherwise it's hard to find/see my reply)

Reply:
I don't care strongly about it either way. Thank you for updating! I'm resolving this conversation.